### ISSUE TITLE
Vagrant isn't F/OSS any more, what to do?

### ISSUE THREAD

**Issue by AdamWill** [fedora-infra/tiny-stage#59]
So, Vagrant [is no longer F/OSS](https://github.com/hashicorp/vagrant/blob/main/LICENSE). Of course Fedora can keep the last F/OSS version around for some time, but this seems like something that might need considering.

There is an [incipient F/OSS fork called Viagrunts](https://github.com/viagrunts/viagrunts) but it doesn't seem to have anything like the momentum of the forks of Terraform or Vault. So I'm not sure we could put all our money on that.

What other options are there? Do it in containers? Any thoughts?

**Comment by abompard** [comment]
Yeah, I followed that as well :-/

I think it would be preferable to avoid containers, because it's much easier to hack things in VMs than in containers, and sometimes I've had to tweak things in the Vagrant VMs when working on other apps that were using them for auth (like changing the ipsilon config and even source code, for example) (yeah sometimes the source code is dangerously close to being a configuration item in Ipsilon, see the OIDC scopes for example).

So, maybe follow with Viagrunts and if something better pops up, move to it, hoping that the configuration remains similar? Most of the work is in the ansible playbooks anyway.

Or maybe the lightweightness of containers overcomes the rare need of editing the source code? It's a developer environment so I think we want to keep it easy to hack, but I'm open to all opinions.

**Comment by AdamWill** [comment]
Well, it's just my opinion, but I've spent the last week and a half building a container-based development environment for Bodhi to replace its VM-based one (actually I've built it three times now...) and so far I *much* prefer it. You can build containers such that it's easy to live-change things in them in various ways; the Bodhi main container uses systemd, so you can shell in, change stuff, and restart services as normal.

The Bodhi app itself is also run via pyramid's `pserve` with `--reload`, which automatically restarts it any time a source file is changed, and is set up to use the live source tree via a volume, so you can just edit code as normal in the host and changes immediately apply in the container. That's only really what you'd want for the specific app you're working on, though, so perhaps less applicable to tinystage.

Even if you don't use systemd within a container there are certainly ways you could set it up so a command can be run inside or outside the container to restart the running server or whatever. It's not an impossible problem. It's not how production containers are usually built, but you can do things differently for a dev environment.

I am in exactly the place you mention on Ipsilon right now trying to work it into the Bodhi dev env, funnily enough - dealing with the OIDC scopes... - and even though the ipsilon container isn't really mutable it's not too bad, as building it is very fast unless the initial package deployment steps have to be re-done. If I get really mad at it it's easy enough just to tweak the container definition to use systemd so I can mess with it live, I may do that today.

So far I definitely find the speed and resource usage improvements with containers heavily outweigh the "convenience" of VMs. I timed it, and bringing up tiny-stage plus the bodhi VM env from scratch takes 39 minutes, and uses around 7G of RAM. I was not able to consistently run the Bodhi test suite in the VM env and use a browser and email client at the same time on my 16G laptop - I'd keep running out of memory and the whole system would choke up unless I ssh'ed in and killed the browser. I really hated it. If I was going to do this, I'd do it with containers.

I do think it's kinda significant that nobody has got behind viagrunts so far; I suspect there just aren't enough people building stuff this way any more cos everyone's doing it in containers. The biggest issue with the viagrunts approach might be making viagrunts an actual thing, I wouldn't want to get saddled with the upstream maintenance. Right now I think it's most just a fork of the vagrant source tree with some superficial renaming done.

**Comment by AdamWill** [comment]
Setting up a container to use systemd is pretty straightforward, btw, it's really just this:

    RUN dnf install -y systemd [along with all your other package installs in a single step, of course]
    COPY your-server.service /etc/systemd/system/your-server.service
    RUN systemctl enable your-server.service
    CMD [ "/usr/sbin/init" ]

...and that's it. If you run that container and then `podman exec -it containername /bin/bash` to get into it, you'll find systemd running, `your-server.service` started, and you can restart it or whatever you like just like normal.

**Comment by abompard** [comment]
Yup, I hadn't investigated this far, sounds like your solution with containers is better indeed.

**Comment by gridhead** [comment]
As of 01 Aug 2024, the namespace on which `viagrunts` had been previously kept has been changed from `viagrunts` to `Be-Secure` and the updated URL to the project is [here](https://github.com/Be-Secure/viagrunts). I am not sure if this is a simple renaming of the namespace or a change of hands but the official [documentation](https://be-secure.github.io/Be-Secure/) does not spark confidence. The [last commit](https://github.com/Be-Secure/viagrunts/commit/115c4a7f05cd65618632adfbfade580992f8169e) to the repository was made on 09 Feb 2024 (which is approximately six months back) and the most [recent tag](https://github.com/Be-Secure/viagrunts/releases/tag/2.4.1.dev%2B000171-55ec0d43) is from 13 Jan 2024 (which is approximately seven months back). The maintainers do not seem to have bothered to change the now-obsolete URL in the about section and the issue tracker seems to be disabled limiting our reach to those who might tell us something.

I have not yet explored the containers-bound implementation but might I suggest `cloud-init`? It is dual licensed under `Apache License 2.0` and `GNU General Public License 3.0 and later` as [mentioned](https://github.com/canonical/cloud-init/blob/main/LICENSE) in their repository. We generate [base cloud images](https://fedoraproject.org/cloud/download) anyway for Fedora Linux and all we need to do is write some YAML instead of Vagrantfile. Sure, this approach would be heavy in terms of storage capacity usage as compared to Vagrant which used to provide us with minimal images. I do think that we should be able to retain most of the `libvirt` based implementation that we have now. The project is well maintained (with the [last commit](https://github.com/canonical/cloud-init/commit/5322dca2f3da7bb9b8d6f1fac6ccb00ef33ef8ee) being made six hours before writing this comment) and the [documentation](https://cloudinit.readthedocs.io/en/latest/) is also extensive.

**Comment by AdamWill** [comment]
Bodhi merged the container-based approach I came up with, and it seems to be working well so far. I'd like to reuse it for tiny-stage but have not had time yet. See https://github.com/fedora-infra/bodhi/blob/develop/docs/developer/bcd.rst for the docs on it.

**Comment by AdamWill** [comment]
Note there is some discussion on this at https://discussion.fedoraproject.org/t/a-future-for-tiny-stage/128580 .
